ENH: Fix output of assert_frame_equal if indexes differ and check_like=True

amilbourne opened the pull request:
- [x] closes #37478 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As described in #37478 This PR fixes some very misleading output when ```assert_frame_equal``` is called with differing index values and ```check_like=True```

I have added a ```check_order``` parameter to the ```assert_index_equal``` function.  This does essentially the same thing as the ```check_like``` parmeter on ```assert_frame_equal``` (but for indexes).  I think ```check_order``` gives a much clearer idea of what the parameter does, but it is a break from the previous naming.  I am unsure whether I should have stuck with ```check_like``` for the new parameter.

amilbourne committed:
Fix output from assert_frame_equal when indexes differ and check_like=True

arw2019 reviewed (commented):
Thanks @amilbourne for the PR! Looks good

I don't know what the pandas policy is on API changes for the testing module (I think it's private but I think it gets used outside our testing suite) so one thing you might need to get agreement on

Related to the above if we regard this as a user-facing change you will need to add a whatsnew note

ivanovmg commented:
> 
> 
> Thanks @amilbourne for the PR! Looks good
> 
> I don't know what the pandas policy is on API changes for the testing module (I think it's private but I think it gets used outside our testing suite) so one thing you might need to get agreement on
> 
> Related to the above if we regard this as a user-facing change you will need to add a whatsnew note

I guess that ``assert_index_equal`` is indeed a part of public API, since the docs are available here https://github.com/pandas-dev/pandas/pull/37479#pullrequestreview-519321722

ivanovmg commented on pandas/_testing.py:
```diff
@@ -667,6 +667,7 @@ def assert_index_equal(
     check_less_precise: Union[bool, int] = no_default,
     check_exact: bool = True,
     check_categorical: bool = True,
+    check_order: bool = True,
     rtol: float = 1.0e-5,
```
Probably it is necessary to have the new parameter added to the very end of this function signature in order not to break any user's code, which for some reason use positional arguments.

amilbourne replied:
Thanks for the feedback.

I debated that and it looked like the ```rtol``` and ```atol``` params had been added before the (older) ```obj``` param, so I decided to put ```check_order``` with the other check params.  That said, ```obj``` is not something that most user code would pass so perhaps the comparrison doesn't hold.

Happy to move the parameter.

I also just noticed that I forgot to add a ```versionadded``` tag (will do).

amilbourne commented:
@ivanovmg, @arw2019  I agree that ```assert_index_equal``` is a part of public API, I just wasn't sure which changes needed a whatsnew note.  I will add one.  Thanks.

amilbourne committed:
Added whatsnew note for new check_order param

amilbourne committed:
Added unit test for new check_order param

amilbourne commented:
I made the suggested changes and added a unit test for the new `check_order` param.

ivanovmg reviewed (approved):
Looks good to me.

amilbourne commented:
The build is failing due to a failing test, but it doesn't seem to be anything to do with the changes I made, so I'm assuming that is an anomaly in the build (or a flakey test or something).

The failing test is:
test_slice_irregular_datetime_index_with_nan from pandas/tests/indexing/test_partial.py

This test is also failing for other unrelated PRs, so I don't think it was me :-)

jreback commented on pandas/_testing.py:
```diff
@@ -707,6 +708,12 @@ def assert_index_equal(
     obj : str, default 'Index'
         Specify object name being compared, internally used to show appropriate
         assertion message.
+    check_order : bool, default True
```
ok to move this after check_categorical

amilbourne replied:
I moved the parameter.

jreback reviewed (changes requested):
this lgtm. just a small comment. pls rebase and ping on green.

amilbourne commented:
@jreback I moved the parameter as asked and rebased.  The build is failing on
`pandas/tests/plotting/test_datetimelike.py::TestTSPlot::test_format_timedelta_ticks_wide`
But I don't think that is related to anything I changed.

jreback commented on pandas/_testing.py:
```diff
@@ -696,6 +697,10 @@ def assert_index_equal(
+    check_order : bool, default True
+        Whether to compare the order of index entries as well as their values.
+        If True, both indexes must contain the same elements, in the same order.
+        If False, both indexes must contain the same elements, but in any order.
```
can you add a versionadded tag 1.2 here

amilbourne replied:
Oh, that's annoying, sorry.
I removed the versionadded tag when I moved the parameters :-(
I have put it back and rebased again.

jreback reviewed (changes requested):
small comment, pls merge master and ping on green.

amilbourne committed:
Adding versionadded tag

jreback commented:
thanks @amilbourne very nice
